pandas/core/indexes/datetimes.py

```diff
@@ -1086,6 +1095,19 @@ def snap(self, freq='S'):
         # we know it conforms; skip check
         return DatetimeIndex(snapped, freq=freq, verify_integrity=False)
 
+    def unique(self, level=None):
+        # Override here since IndexOpsMixin.unique uses self._values.unique
+        # For DatetimeIndex with TZ, that's a DatetimeIndex -> recursion error
+        # So we extract the tz-naive DatetimeIndex, unique that, and wrap the
+        # result with out TZ.
+        if self.tz is not None:
+            naive = type(self)(self._ndarray_values, copy=False)
```

Comment:
huh?

Reply:
Do you have a concrete question? It's not the nicest design, but for me the comment above is clear (and I suppose the idea is that this is temporary until there is a DatetimeTZArray)

Reply:
all of this hacking around is not great. The problem is these temporary work-around become more and more. Really need to bite the bullet and fix up the implementations to avoid this.